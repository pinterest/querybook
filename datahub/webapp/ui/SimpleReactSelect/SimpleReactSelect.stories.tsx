import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { SimpleReactSelect } from './SimpleReactSelect';

export const _SimpleReactSelect = (args) => {
    const [value, setValue] = React.useState(1);

    return (
        <div style={{ width: '160px', alignItems: 'stretch' }}>
            <SimpleReactSelect
                {...args}
                value={value}
                onChange={setValue}
                options={[
                    { label: 'option one', value: 1 },
                    { label: 'option two', value: 2 },
                ]}
            />
        </div>
    );
};

_SimpleReactSelect.args = {
    withDeselect: false,
    isDisabled: false,
    clearAfterSelect: false,
};

export default {
    title: 'Form/SimpleReactSelect',
    decorators: [centered],
};
