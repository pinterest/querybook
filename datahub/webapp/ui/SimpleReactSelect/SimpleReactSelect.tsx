import React, { useMemo, useCallback } from 'react';
import { Props as ReactSelectProps } from 'react-select/lib/Select';
import Select from 'react-select';

import { makeReactSelectStyle } from 'lib/utils/react-select';

interface ISelectOption<T> {
    value: T;
    label: string;
}

export interface ISimpleReactSelectProps<T> {
    options: Array<ISelectOption<T>>;
    value: T;
    onChange: (o: T) => any;
    withDeselect?: boolean;
    isDisabled?: boolean;
    selectProps?: Partial<ReactSelectProps<T>>;
}

const reactSelectStyle = makeReactSelectStyle(true);

export function SimpleReactSelect<T>({
    options,
    value,
    onChange,
    withDeselect = false,
    isDisabled,
    selectProps = {},
}: ISimpleReactSelectProps<T>) {
    const selectedOption = useMemo(
        () => options.find((option) => option.value === value),
        [options, value]
    );
    const computedOptions = useMemo(
        () =>
            withDeselect
                ? [{ value: undefined, label: 'Deselect' }].concat(options)
                : options,
        [withDeselect, options]
    );
    const onSelectChange = useCallback(
        (val: ISelectOption<T>) => onChange(val?.value),
        [onChange]
    );

    return (
        <Select
            styles={reactSelectStyle}
            menuPortalTarget={document.body}
            value={selectedOption}
            onChange={onSelectChange}
            options={computedOptions}
            isDisabled={isDisabled}
            {...selectProps}
        />
    );
}
