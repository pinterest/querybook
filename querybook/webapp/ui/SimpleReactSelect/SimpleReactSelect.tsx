import React, { useCallback, useMemo } from 'react';
import Select, { Props as ReactSelectProps } from 'react-select';
import Creatable from 'react-select/creatable';

import { makeReactSelectStyle } from 'lib/utils/react-select';
import { overlayRoot } from 'ui/Overlay/Overlay';
import { AccentText } from 'ui/StyledText/StyledText';

export interface ISelectOption<T> {
    value: T;
    label: React.ReactNode;
}

export interface ISimpleReactSelectProps<T> {
    options: Array<ISelectOption<T> | string>;
    value: T;
    onChange: (o: T) => any;
    withDeselect?: boolean;
    isDisabled?: boolean;
    creatable?: boolean;
    selectProps?: Partial<ReactSelectProps<T>>;

    // Clear selection user picks value
    clearAfterSelect?: boolean;
}

const reactSelectStyle = makeReactSelectStyle(true);

export function SimpleReactSelect<T>({
    options,
    value,
    onChange,
    isDisabled,
    creatable,

    selectProps = {},
    withDeselect = false,
    clearAfterSelect = false,
}: ISimpleReactSelectProps<T>) {
    const overrideSelectProps = useMemo(() => {
        const override: Partial<ReactSelectProps<T>> = {};
        if (clearAfterSelect) {
            override.value = null;
        }

        return override;
    }, [clearAfterSelect]);

    const computedOptions = useMemo(
        () =>
            (options || []).map((option) =>
                typeof option === 'string'
                    ? {
                          label: option,
                          value: option,
                      }
                    : option
            ),
        [options]
    );

    const selectedOption = useMemo(
        () => computedOptions.find((option) => option.value === value),
        [computedOptions, value]
    );

    const onSelectChange = useCallback(
        (val: ISelectOption<T>) => onChange(val?.value),
        [onChange]
    );

    const componentProps = {
        styles: reactSelectStyle,
        menuPortalTarget: overlayRoot,
        value: selectedOption,
        onChange: onSelectChange,
        options: computedOptions,
        isDisabled,
        isClearable: withDeselect,
        menuPlacement: 'auto' as const,
        placeholder: 'Select',
        ...selectProps,
        ...overrideSelectProps,
    };

    return (
        <AccentText>
            {creatable ? (
                <Creatable {...componentProps} />
            ) : (
                <Select {...componentProps} />
            )}
        </AccentText>
    );
}
