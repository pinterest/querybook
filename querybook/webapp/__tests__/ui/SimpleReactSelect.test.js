import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { SimpleReactSelect } from '../../ui/SimpleReactSelect/SimpleReactSelect';

it('renders without crashing', () => {
    shallow(
        <SimpleReactSelect options={[]} value="test" onChange={() => null} />
    );
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(
            <SimpleReactSelect
                options={[]}
                value="test"
                onChange={() => null}
            />
        );
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(
            <SimpleReactSelect
                options={[]}
                value="test"
                onChange={() => null}
            />
        );
        expect(output).toMatchSnapshot();
    });
});
