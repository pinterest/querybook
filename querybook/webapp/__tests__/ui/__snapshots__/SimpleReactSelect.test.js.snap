// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<AccentText>
  <StateManager
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isClearable={false}
    menuPlacement="auto"
    menuPortalTarget={<div />}
    onChange={[Function]}
    options={Array []}
    placeholder="Select"
    styles={
      Object {
        "clearIndicator": [Function],
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "input": [Function],
        "menu": [Function],
        "menuList": [Function],
        "menuPortal": [Function],
        "multiValue": [Function],
        "multiValueLabel": [Function],
        "multiValueRemove": [Function],
        "option": [Function],
        "placeholder": [Function],
        "singleValue": [Function],
      }
    }
  />
</AccentText>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<div
  className="StyledText__StyledTextDiv-sc-1ldibmc-0 bLpBLx Text"
>
  <div
    className=" css-2b097c-container"
    onKeyDown={[Function]}
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className=" css-1klw407-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className=" css-l1xr2i-Component"
      >
        <div
          className=" css-phc9i8-placeholder"
        >
          Select
        </div>
        <div
          className="css-4glw1s-Component"
        >
          <div
            className=""
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="react-select-2-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "label": "input",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className=" css-ny0e4k-Component"
      >
        <span
          className=" css-shgobz-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className=" css-1mkmlcg-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
